src/lib.rs

```diff
@@ -135,12 +135,53 @@ fn squared_diff_norm(
     sum
 }
 
+fn apply_accum_helper<const Cross: bool>(
+    accum: &mut Mean,
+    points_a: &PointProps,
+    points_b: &PointProps,
+    squared_bin_edges: &[f64],
+) {
+    for i_a in 0..points_a.n_points {
+        let i_b_start = if Cross { i_a + 1 } else { 0 };
+        for i_b in i_b_start..points_b.n_points {
+            // compute the distance between the points, then the distance bin
+            let distance_squared = squared_diff_norm(
+                points_a.positions,
+                points_b.positions,
+                i_a,
+                i_b,
+                points_a.n_points,
+                points_b.n_points,
+                points_a.n_spatial_dims,
+            );
+            if let Some(distance_bin_idx) = get_distance_bin(distance_squared, &squared_bin_edges) {
+                // get the value. This is hardcoded to correspont to the velocity structure
+                // function when accumulated with Mean
+                // TODO switch on pairwise op?
+                let val = squared_diff_norm(
+                    &points_a.values,
+                    &points_b.values,
+                    i_a,
+                    i_b,
+                    points_a.n_points,
+                    points_b.n_points,
+                    points_a.n_spatial_dims,
+                )
+                .sqrt();
+
+                // get the weight
+                let pair_weight = points_a.get_weight(i_a) * points_b.get_weight(i_b);
+
+                accum.consume(val, pair_weight, distance_bin_idx);
+            }
+        }
+    }
+}
+
 // maybe we want to make separate functions for auto-stats vs
 // cross-stats
 // TODO: generalize to allow faster calculations for regular spatial grids
 pub fn apply_accum(
-    out: &mut Vec<f64>,
-    weights_out: &mut Vec<f64>,
```

Comment:
I feel pretty strongly that we don't want to return this. I think we want people to directly call `accum.get_value(...)`.

Reply:
I'm happy to elaborate more

Reply:
No, that makes sense.